src/controllers/AuthenticationController.ts

```diff
@@ -0,0 +1,345 @@
+      }
+
+      try {
+        JWTUtils.verifyAccessToken(token, user.password);
+      } catch (e) {
+        res.status(401);
+        res.json({message: userFriendlyMessage.failure.invalidToken});
+        return;
+      }
+
+      const {password, passwordConfirmation} = req.body;
+
+      if (password !== passwordConfirmation) {
+        res.status(400);
+        res.json({
+          message: userFriendlyMessage.failure.passwordConfirmationMismatch,
+        });
+        return;
+      }
+
+      const updatedAttributes: UserAttributes = {
+        ...user,
+        password: password,
+        isConfirmed: true,
+        isTemporary: false,
+      };
+      await this.userService.updateOneUserById(id, updatedAttributes);
+      res.json({message: userFriendlyMessage.success.setPassword});
+      // Redirect user to login page
+    } catch (e) {
+      res.status(400);
+      res.json({message: userFriendlyMessage.failure.setPassword});
+      next(e);
+    }
+  }
+
+  async resetPassword(req: Request, res: Response, next: NextFunction) {
+    try {
+      //TODO: Auth Middleware will add the current user in req.user
+      const {user} = req;
+      const {originalPassword, newPassword, newPasswordConfirmation} = req.body;
+
+      if (!user) {
+        res.status(400);
+        res.json({message: userFriendlyMessage.failure.userNotExist});
+        return;
+      }
+      if (!(await user.isPasswordMatch(originalPassword))) {
+        res.status(401);
+        res.json({message: userFriendlyMessage.failure.incorrectPassword});
+        return;
+      }
+      if (newPassword !== newPasswordConfirmation) {
+        res.status(400);
+        res.json({
+          message: userFriendlyMessage.failure.passwordConfirmationMismatch,
+        });
+        return;
+      }
+
+      const id = user.id;
+      const updatedAttributes: UserAttributes = {
+        ...user,
+        password: newPassword,
+      };
+      const updatedUser = await this.userService.updateOneUserById(
+        id,
+        updatedAttributes
+      );
+      res.json({
+        message: userFriendlyMessage.success.resetPassword,
+        data: updatedUser,
+      });
+    } catch (e) {
+      res.status(400);
+      res.json({message: userFriendlyMessage.failure.resetPassword});
+      next(e);
+    }
+  }
+
+  async forgetPasswordEmail(req: Request, res: Response, next: NextFunction) {
+    try {
+      const {email} = req.body;
+      const user = await this.userService.getOneUserByEmail(email);
+
+      if (!user) {
+        res.status(400);
+        res.json({message: userFriendlyMessage.failure.emailNotExist});
+        return;
+      }
+
+      const payload: Payload = {
+        id: user.id,
+        role: user.role,
+        schoolId: user.schoolId,
+        isConfirmed: user.isConfirmed,
+        isTemporary: user.isTemporary,
+      };
+      const accessToken = JWTUtils.generateAccessToken(
+        payload,
+        user.password,
+        {}
```

Comment:
Forget Password should expire in 30 minutes

Reply:
Forget password's token should also be a new unique token used ONLY for forget password, it shouldn't be the same as the normal ones. Else people can abuse, click forget password get a token for free without signing in and have access to all our APIs.

Reply:
As discussed, will use user's old password as secret, so that the set password link will only work once. Will reuse the controller method for `setPassword` with another route for `setForgetPassword`.